### ISSUE TITLE
Feature: async

### ISSUE THREAD

**Issue by mfbx9da4** [mrousavy/react-native-mmkv#446]
While mmkv is super fast, and that's awesome 😎 it still blocks the main JS thread and for situations where you have to get a bunch of keys in a loop of some kind, that can add up. It would be great if mmkv could provide a threaded async version where
* Stored values are loaded from mmkv in a thread 
* JSI strings and numbers are constructed in a thread
* Promise / callback is resolved back to main JS thread with `jsCallInvoker`

Do you think that would be possible?

**Comment by mrousavy** [comment]
Yep, that's definitely possible!

I'm not sure if I'd be happy adding an async method for every sync method though, because the API will look a bit messy then. Right now it's very simple and minimal. 😅 Maybe we can find a way around that

Either way, it's a little bit of work, not sure how much exactly but we need to set up a Promise resolver that runs on a separate Thread, custom batched methods, etc etc

**Comment by mrousavy** [comment]
Arg parsing will need to happen outside, but that's fast anyways.

**Comment by mfbx9da4** [comment]
🤔 

Perhaps batch would have to return an array of items, does that solve it?

```typescript
const storage = new MMKV()
const [foo] = await storage.batch((s) => [
  s.getString('foo'),
  s.set('hello', 'world'),
  s.set('test', 'another')
])
```

**Comment by mfbx9da4** [comment]
Also could you not make storage to be context aware provided you enforce that batch is always a synchronous function and then you could do 

```typescript
const storage = new MMKV()
const [foo] = await storage.batch(() => [
  storage.getString('foo'),
  storage.set('hello', 'world'),
  storage.set('test', 'another')
])
```

BTW, I love the batch() idea

**Comment by mrousavy** [comment]
> Perhaps batch would have to return an array of items

That's a good idea, but then you'd be very limited in terms of doing stuff contextually (e.g. only set if the value in storage is ....), but yea that would probably work.

> Provided you enforce that batch is always a synchronous function.

Wdym? The whole idea is to have batch asynchronous and awaitable

**Comment by mfbx9da4** [comment]
> Wdym? The whole idea is to have batch asynchronous and awaitable

Sorry I mean the batch implementation would enforce that the batch argument callback is synchronous

```typescript
class Storage {
  // ...
  batch(fn) {
    ensureSynchronous(fn)
    this._batchingEnabled = true
    try {
      fn()
    } finally {
      this._batchingEnabled = false
    }
  }
}
```

**Comment by mfbx9da4** [comment]
> but then you'd be very limited in terms of doing stuff contextually (e.g. only set if the value in storage is ....)

What use case might be limited?

**Comment by mrousavy** [comment]
eg. this (but on a larger scale with tons of items like arrays):
```ts
const storage = new MMKV()
await storage.batch((s) => {
  const x = s.getBoolean('bla')
  if (x === true) {
    s.set('hello', 'world') 
  } else {
    s.set('test', 'another')
  }
  // ...
})
```

**Comment by mfbx9da4** [comment]
Hmm yeah it's true that wouldn't work and having a callback does give the impression that it should be possible. One saving grace is that typescript users will get the following error:

```typescript
const storage = new MMKV()
await storage.batch((s) => {
  const x = s.getBoolean('bla')
  // ❌ This condition will always return 'false' since the types 'MMKVBatchItem' and 'boolean' have no overlap
  if (x === true) {
    s.set('hello', 'world') 
  } else {
    s.set('test', 'another')
  }
  // ...
})
```

I don't see how you could support the above API, unless... you implemented a similar strategy to react-native-multithreading and all batched functions were marked by default with a "worklet" tag?

**Comment by mfbx9da4** [comment]
Also maybe just twitter poll the options?

**Comment by elliotsayes** [comment]
+1 For async, as it will make larger buffer storage play nicely :)

In the meantime, is there any way to wrap the existing methods so they don't block the main thread ?

**Comment by ex3ndr** [comment]
We have a project where mmkv overgrown it and currently blocks often main thread

**Comment by mrousavy** [comment]
Yea I think the async batch methods are cool, there are some things we need to think about though. I also need a lot of free time to work on this, so if you want to see this happening consider sponsoring me / emailing me about those efforts

**Comment by bfricka** [comment]
> +1 For async, as it will make larger buffer storage play nicely :)
> 
> In the meantime, is there any way to wrap the existing methods so they don't block the main thread ?

Not really, if the underlying implementation is synchronous. The best you could do would be to defer execution to the next tick as a microtask, using promises, requestAnimationFrame, timeout, etc. E.g. 

```ts
const resolvedPromise = Promise.resolve()
const setAsync = (k: string, v: string) => resolvedPromise.then(() => storage.set(k, v))
setAsync('foo', 'bar')
```

This would defer to the next tick, running before any timeouts, but it would still block. Async in this way is only useful if your underlying implementation eventually uses threads and doesn't block the main thread. Otherwise, you're just blocking later instead of now.

**Comment by emzet93** [comment]
I was wondering if it would be possible to delegate saving to MMKV to [Worklet](https://github.com/margelo/react-native-worklets-core), so it doesn't block the main thread? I was trying to implement that but I'm facing issues with executing mmkv instance methods within the worklet. And I'm actually worried that the reason is that I don't really understand the concept of worklets properly 😅

**Comment by XantreGodlike** [comment]
This shouldn't be implemented on worklets, because data sharing with other thread is a big overhead

**Comment by emzet93** [comment]
that makes sense, thanks for explanation @XantreGodlike!

**Comment by JustJoostNL** [comment]
Is this feature planned for in the future?

**Comment by vilavren** [comment]
Is this feature planned for in the future?

**Comment by mrousavy** [comment]
Maybe when I rewrite the library to [Nitro](https://nitro.margelo.com). But please stop spamming.

**Comment by vilavren** [comment]
Thank you very much, we are looking forward to it!

**Comment by mrousavy** [comment]
Hey - since react-native-mmkv is now on Nitro, you can already achieve this by simply using it in a [Worklet](https://docs.swmansion.com/react-native-worklets/docs/) - since all [Nitro Modules](https://nitro.margelo.com) can be used inside Worklets!

```ts
const storage = createMMKV()
const runtime = createWorkletRuntime({ name: 'background' })
const result = await runOnRuntimeAsync(runtime, () => {
  'worklet'
  storage.set('heavy-key', 'heavy value goes here')
  return storage.getString('heavy-key')
})
```

**Comment by mrousavy** [comment]
It's not exactly the same as asynchronous `set` and `get` methods, but I think this is still super good - and doesnt require me to 2x my test surface in MMKV :) 

Let me know if you guys have any issues with this, maybe there is a reason to add async `set` and `get` methods after all. Otherwise; closing.

**Comment by AndreiCalazans** [comment]
The problem is worklets aren't performant, their serializing step is synchronous on the JS thread. 

<img width="300"  alt="Image" src="https://github.com/user-attachments/assets/c79e8bdc-0956-43b6-9a59-df26ff12a52e" />


Using the Worklet runtime will only be feasible for small data which would just be worth using MMKV's synchronous APis for.
